java/source/infodynamics/measures/mixed/kraskov/MutualInfoCalculatorMultiVariateWithDiscreteKraskov.java

```diff
@@ -247,124 +461,130 @@ protected void computeNorms() {
-			discreteData = MatrixUtils.extractSelectedTimePoints(discreteData, reordering);
-			// Compute the MI
-			double newMI = computeAverageLocalOfObservationsWhileComputingDistances();
-			// restore data2
-			discreteData = originalDiscreteData;
-			return newMI;
-		}
+	public double computeAverageLocalOfObservations(int[] newOrdering)
+			throws Exception {
 		
-		// Otherwise we will use the norms we've already computed, and use a "virtual"
-		//  reordered data2.
-		int[] reorderedDiscreteData = MatrixUtils.extractSelectedTimePoints(discreteData, reordering);
-
-		if (xNorms == null) {
-			computeNorms();
+		if (newOrdering == null) {
+			return computeAverageLocalOfObservations();
 		}
 
-		// Count the average number of points within eps_x and eps_y
-		double averageDiGammas = 0;
-		double avNx = 0;
-		double avNy = 0;
-		
-		for (int t = 0; t < N; t++) {
-			// Compute eps_x and eps_y for this time step:
-			//  using x norms to all neighbours
-			//  (note that norm of point t to itself will be set to infinity).
-			// Then find the k closest neighbours in the same discrete bin
-			double eps_x = MatrixUtils.kthMinSubjectTo(xNorms[t], k, reorderedDiscreteData, reorderedDiscreteData[t]);			
-			
-			// Count the number of points whose x distance is less
-			//  than or equal to eps_x
-			int n_x = 0;
-			for (int t2 = 0; t2 < N; t2++) {
-				if (xNorms[t][t2] <= eps_x) {
-					n_x++;
-				}
-			}
-			// n_y is number of points in that bin minus that point
-			int n_y = counts[reorderedDiscreteData[t]] - 1;
-			avNx += n_x;
-			avNy += n_y;
-			// And take the digamma before adding into the 
-			//  average:
-			averageDiGammas += MathsUtils.digamma(n_x) + MathsUtils.digamma(n_y);
-		}
-		averageDiGammas /= (double) N;
-		if (debug) {
-			avNx /= (double)N;
-			avNy /= (double)N;
-			System.out.println(String.format("Average n_x=%.3f, Average n_y=%.3f", avNx, avNy));
-		}
+		// Take a clone of the object to compute the MI of the surrogates:
+		// (this is a shallow copy, it doesn't make new copies of all
+		//  the arrays)
+		MutualInfoCalculatorMultiVariateWithDiscreteKraskov miSurrogateCalculator =
+				(MutualInfoCalculatorMultiVariateWithDiscreteKraskov) this.clone();
 		
-		// Don't need the 1/k correction here because the conditional entropy term
-		//  is taken over the continuous space only. The correction is (m-1)/k
-		//  for an entropy over m subspaces.
-		// mi = MathsUtils.digamma(k) - 1.0/(double)k - averageDiGammas + MathsUtils.digamma(N);
-		// Instead do:
-		mi = MathsUtils.digamma(k) - averageDiGammas + MathsUtils.digamma(N);
-		miComputed = true;
-		return mi;
+		// Generate a new re-ordered data set
+		int[] shuffledDiscreteData = MatrixUtils.extractSelectedTimePoints(discreteData, newOrdering);
+		// Perform new initialisations
+		miSurrogateCalculator.initialise(dimensions, base);
+		// Set new observations
+		miSurrogateCalculator.setObservations(continuousData, shuffledDiscreteData);

```

Comment:
Here if time_diff != 0 the clone will still have the time_diff set, so then when you setObservations with the continuousData and shuffledDiscreteData, it will try to use the time_diff again (a second time) when pulling the data together for an MI calculation. Now, since we're computing surrogates, it doesn't matter *so* much, but it will use abs(timeDiff) less samples than it should do. I'm being pedantic, but tor 100% integrity could you please set time_diff to 0 on the clone? I have a feeling this code is copied and adjusted from somewhere else and that I should change it there too, so please do let me know if you remember (it doesn't seem to be the original KSG calculator)

Reply:
Good catch! Fixed in https://github.com/jlizier/jidt/pull/58/commits/8f169171e110752185f4766d2f1a398e07dfa131